src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts

```diff
@@ -115,6 +118,92 @@ class RemoteMachineTrainingService implements TrainingService {
+        const deferred: Deferred<void> = new Deferred<void>();
+        if(!trial.rmMeta) {
+            throw new Error(`rmMeta not set in trial ${trial.id}`);
+        }
+        let sshClientManager: SSHClientManager | undefined = this.machineSSHClientMap.get(trial.rmMeta);
+        if(!sshClientManager) {
+            throw new Error(`remoteSSHClient not initialized`);
+        }
+        let sshClient: Client | undefined = sshClientManager.getAvailableSSHClient();
+        if(!sshClient) {
+            //create a new connection
+            await this.createSSHClient(trial);
+        }else {
+            this.trialSSHClientMap.set(trial.id, sshClient);
+        }
+
+        deferred.resolve();
+        return deferred.promise;
+    }
+
+    /**
+     * Create a new ssh connection client by trial job id
+     */
+    public async createSSHClient(trial: RemoteMachineTrialJobDetail): Promise<void> {
+        const deferred: Deferred<void> = new Deferred<void>();
+        const rmMeta = trial.rmMeta;
+        if (rmMeta === undefined) {
+            throw new Error(`rmMeta not set for ${trial.id}`);
+        }
+        const conn: Client = new Client();
+        let connectConfig: ConnectConfig = {
+            host: rmMeta.ip,
+            port: rmMeta.port,
+            username: rmMeta.username };
+        if (rmMeta.passwd) {
+            connectConfig.password = rmMeta.passwd;                
+        } else if(rmMeta.sshKeyPath) {
+            if(!fs.existsSync(rmMeta.sshKeyPath)) {
+                //SSh key path is not a valid file, reject
+                deferred.reject(new Error(`${rmMeta.sshKeyPath} does not exist.`));
+            }
+            const privateKey: string = fs.readFileSync(rmMeta.sshKeyPath, 'utf8');
+
+            connectConfig.privateKey = privateKey;
+            connectConfig.passphrase = rmMeta.passphrase;
+        } else {
+            deferred.reject(new Error(`No valid passwd or sshKeyPath is configed.`));
+        }
+        conn.on('ready', () => {
+            this.trialSSHClientMap.set(trial.id, conn);
+            let sshClientManager: SSHClientManager | undefined = this.machineSSHClientMap.get(rmMeta);
+            if(!sshClientManager) {
+                throw new Error(`ssh clientManager is not initialized in trial ${trial.id}!`);
+            }
+            sshClientManager.addNewSSHClient(conn);
+            this.machineSSHClientMap.set(rmMeta, sshClientManager);
+            deferred.resolve();
+        }).on('error', (err: Error) => {
+            // SSH connection error, reject with error message
+            deferred.reject(new Error(err.message));
+        }).connect(connectConfig);
+      
+        return deferred.promise;
+    }
+    
+    /**
+     * If a trial is finished, retrieve the connection resource
+     * @param trial 
+     */
+    public retrieveSSHClient(trial: RemoteMachineTrialJobDetail): void {
```

Comment:
The naming of this function is confused.

Reply:
changed to `releaseTrialSSHClient`